src/ray/core_worker/transport/direct_actor_transport.h

```diff
@@ -125,60 +126,139 @@ class CoreWorkerDirectActorTaskSubmitter {
   friend class CoreWorkerTest;
 };
 
+/// Object dependency and RPC state of an inbound request.
+class InboundRequest {
+ public:
+  InboundRequest(){};
+  InboundRequest(std::function<void()> accept_callback,
+                 std::function<void()> reject_callback, bool has_dependencies)
+      : accept_callback_(accept_callback),
+        reject_callback_(reject_callback),
+        has_pending_dependencies_(has_dependencies) {}
+
+  void Accept() { accept_callback_(); }
+  void Cancel() { reject_callback_(); }
+  bool CanExecute() { return !has_pending_dependencies_; }
+  void MarkDependenciesSatisfied() { has_pending_dependencies_ = false; }
+
+ private:
+  std::function<void()> accept_callback_;
+  std::function<void()> reject_callback_;
+  bool has_pending_dependencies_;
+};
+
+/// Waits for an object dependency to become available. Abstract for testing.
+class DependencyWaiter {
+ public:
+  /// Calls `callback` once the specified objects become available.
+  virtual void Wait(const std::vector<ObjectID> &dependencies,
+                    std::function<void()> on_dependencies_available) = 0;
+};
+
+class DependencyWaiterImpl : public DependencyWaiter {
+ public:
+  DependencyWaiterImpl(RayletClient &raylet_client) : raylet_client_(raylet_client) {}
+
+  void Wait(const std::vector<ObjectID> &dependencies,
+            std::function<void()> on_dependencies_available) override {
+    auto tag = next_request_id_++;
+    requests_[tag] = on_dependencies_available;
+    raylet_client_.WaitForDirectActorCallArgs(dependencies, tag);
+  }
+
+  /// Fulfills the callback stored by Wait().
+  void OnWaitComplete(int64_t tag) {
```

Comment:
This seems like it should also be part of the interface.

Reply:
It's an implementation detail of this particular one, the mock doesn't require it.